devicetree: Add clock-names support

galak opened the pull request.

galak committed:
devicetree.h: clocks: Add support for _BY_NAME macros

Add the following macro's to get clock info by name:
	DT_CLOCKS_LABEL_BY_NAME
	DT_CLOCKS_CELL_BY_NAME
	DT_INST_CLOCKS_LABEL_BY_NAME
	DT_INST_CLOCKS_CELL_BY_NAME

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Co-Authored-By: Marti Bolivar <marti.bolivar@nordicsemi.no>

galak committed:
dts/bindings: Add clock-names as a supported property

We've had clocks in base.yaml but didn't have clock-names.  Add it to
base.yaml with similar functionality to interrupt-names, reg-names, etc.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -54,6 +54,39 @@ extern "C" {
+ *
+ * Example devicetree fragment:
+ *
+ *     clk1: clkctrl@... {
```
I took a look, and it seems like dtspec recommends either 'clock' or 'clock-controller' for these. Think it's worth changing the examples to align with the spec as follow up work? I know this is just using the existing convention so it's fine for me to keep it as is for this PR.

galak replied:
Added a commit in this PR for it.

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -54,6 +54,39 @@ extern "C" {
+ * @see DT_PHANDLE_BY_NAME()
+ */
+#define DT_CLOCKS_LABEL_BY_NAME(node_id, name) \
+	DT_PROP(DT_PHANDLE_BY_NAME(node_id, clocks, name), label)
```
You can just use DT_LABEL(DT_PHANDLE_BY_NAME(...)) if you prefer.

galak replied:
Was matching what we had in dma.h and adc.h.

mbolivar-nordic replied:
Ok by me to defer to future cleanup

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -96,6 +129,41 @@ extern "C" {
 	DT_PHA_BY_IDX(node_id, clocks, idx, cell)
 
+/**
+ * @brief Get Clock controller "cell" value by name
```
```suggestion
 * @brief Get clock controller "cell" value by name
```

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -96,6 +129,41 @@ extern "C" {
+ *     DT_CLOCKS_CELL_BY_NAME(DT_NODELABEL(n), beta, bits) // 40
+ *
+ * @param node_id node identifier
+ * @param cell binding's cell name within the specifier referenced as "name"
```
Doxygen has cell, then name, but the macro has name, then cell, as the parameter order.

galak replied:
fixed.

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -96,6 +129,41 @@ extern "C" {
+ * @param node_id node identifier
+ * @param cell binding's cell name within the specifier referenced as "name"
+ * @param name lowercase-and-underscores specifier name
+ * @return the value of the cell inside the named specifier by name
```
```suggestion
 * @return the value of the cell inside the named specifier
```

Just a bit of redundancy there

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -115,6 +183,16 @@ extern "C" {
 	DT_CLOCKS_LABEL_BY_IDX(DT_DRV_INST(inst), idx)
 
+/**
+ * @brief Get a DT_DRV_COMPAT clock controller "name" by name
```
```suggestion
 * @brief Get a DT_DRV_COMPAT clock controller "name" (label property) by name
```

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -133,6 +211,16 @@ extern "C" {
 
+/**
+ * @brief Get a DT_DRV_COMPAT clock controller "cell" value by name
+ * @param inst instance number
```
```suggestion
 * (see @ref DT_CLOCKS_CELL_BY_NAME)
 * @param inst instance number
```

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -133,6 +211,16 @@ extern "C" {
+ * @param inst instance number
+ * @param name lowercase-and-underscores specifier name
+ * @param cell binding's cell name within the specifier referenced as "name"
+ * @return the value of the cell inside the named specifier by name
```
```suggestion
 * @return the value of the cell inside the named specifier
```

mbolivar-nordic commented on tests/lib/devicetree/src/main.c:
```diff
@@ -1173,6 +1173,11 @@ static void test_clocks(void)
+	/* DT_CLOCKS_LABEL_BY_NAME */
+	zassert_true(!strcmp(DT_CLOCKS_LABEL_BY_NAME(TEST_TEMP, clk_b),
+			     "TEST_CLOCK"),
+		     "label fixed");
```
```suggestion
		     "label b");
```

mbolivar-nordic commented on tests/lib/devicetree/src/main.c:
```diff
@@ -1182,6 +1187,12 @@ static void test_clocks(void)
+	zassert_equal(DT_CLOCKS_CELL_BY_NAME(TEST_TEMP, clk_a, bits), 7,
+		      "clk-a bits");
+	zassert_equal(DT_CLOCKS_CELL_BY_NAME(TEST_TEMP, clk_b, bus), 8,
+		      "clk-a bus");
```
```suggestion
		      "clk-b bus");
```

mbolivar-nordic commented on tests/lib/devicetree/src/main.c:
```diff
@@ -1199,6 +1210,11 @@ static void test_clocks(void)
+	/* DT_INST_CLOCKS_LABEL_BY_NAME */
+	zassert_true(!strcmp(DT_INST_CLOCKS_LABEL_BY_NAME(0, clk_b),
+			     "TEST_CLOCK"),
+		     "label fixed");
```
```suggestion
		     "label b");
```

mbolivar-nordic commented on tests/lib/devicetree/src/main.c:
```diff
@@ -1208,6 +1224,12 @@ static void test_clocks(void)
+	zassert_equal(DT_INST_CLOCKS_CELL_BY_NAME(0, clk_a, bits), 7,
+		      "clk-a bits");
+	zassert_equal(DT_INST_CLOCKS_CELL_BY_NAME(0, clk_b, bus), 8,
+		      "clk-a bus");
```
```suggestion
		      "clk-b bus");
```

galak committed:
devicetree.h: clocks: Rename clock node name

Rename clock controller node name to match recommend generic name from
device tree spec.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
